File: java/src/main/java/Say.java
import java.util.List;

public class Say {
  /**
   * Words representing 10s.
   */
  static final List<String> TENS = List.of(
          "", "ten", "twenty", "thirty", "forty", "fifty", "sixty", "seventy", "eighty", "ninety"
  );

  /**
   * Zero text.
   */
  private static final String ZERO_TEXT = "zero";
    
  /**
   * Words representing units.
   */
  private static final List<String> UNITS = List.of("", "one", "two", "three", "four", "five", "six", "seven", "eight",
                                                    "nine");
    
  /**
   * Words representing numbers between 10 and 19.
   */
  private static final List<String> TEENS = List.of(
          "ten", "eleven", "twelve", "thirteen", "fourteen", "fifteen", "sixteen", "seventeen", "eighteen", "nineteen"
  );
    

  public String say(final long number) {
    if (number < 0) throw new IllegalArgumentException("number must be non-negative");
    if (number >= 1_000_000_000_000L) throw new IllegalArgumentException("number too large");
    String result = "";  
    if (number == 0) {
        result = "zero";
    } else {
        int subThousand = (int)(number % 1000);
        int thousands = (int)(number / 1000 % 1000);
        int millions = (int)(number / 1_000_000 % 1000);
        int billions = (int)(number / 1_000_000_000 % 1000);
        
        if (billions > 0) {
            result = result + getSubThousandText(billions) + " billion ";
        }
        if (millions > 0) {
            result = result + getSubThousandText(millions) + " million ";
        }
        if (thousands > 0) {
            result = result + getSubThousandText(thousands) + " thousand ";
        }
        if (subThousand > 0) {
            result = result + getSubThousandText(subThousand);
        }
    }
    return result.trim();
  }

  /**
   * Determine text for numbers between 1 and 999.
   *
   * @param subThousand Number to convert - must be between 1 and 999.
   * @return converted number.
   */
  private String getSubThousandText(final int subThousand) {
    final StringBuilder result = new StringBuilder();
    final int subHundred = subThousand % 100;
    if (subThousand >= 100) {
      result.append(getSubHundredText(subThousand / 100))
            .append(" ")
            .append("hundred");
      if (subHundred > 0) {
        result.append(" ");
      }
    }
    if (subHundred > 0) {
      result.append(getSubHundredText(subHundred));
    }
    return result.toString();
  }

  /**
   * Determine text for numbers between 1 and 99.
   *
   * @param subHundred Number to convert - must be between 1 and 99.
   * @return converted number.
   */
  private String getSubHundredText(final int subHundred) {
    final String result;
    if (subHundred >= 10 && subHundred < 20) {
      result = TEENS.get(subHundred - 10);
    } else {
      final int tens = subHundred / 10;
      final String tensString = TENS.get(tens);

      final int units = subHundred % 10;
      final String unitsString = UNITS.get(units);

      if (tens == 0) {
        result = unitsString;
      } else if (units == 0) {
        result = tensString;
      } else {
        result = tensString + "-" + unitsString;
      }
    }
    return result;
  }
}

File: scala/src/main/scala/Say.scala
import scala.collection.immutable.ListMap

object Say {
  private val Units = List("one", "two", "three", "four", "five", "six", "seven", "eight", "nine")
  private val Teens = List("eleven", "twelve", "thirteen", "fourteen", "fifteen", "sixteen", "seventeen", "eighteen", "nineteen")
  private val Tens = List("ten", "twenty", "thirty", "forty", "fifty", "sixty", "seventy", "eighty", "ninety")
  private val Triplets = ListMap(
    1000_000_000 -> Some("billion"),
    1000_000 -> Some("million"),
    1000 -> Some("thousand"),
    1 -> None,
  )

  private def inEnglishSubHundred(number: Int): String = (number / 10, number % 10) match {
    case (0, u) => Units(u - 1)
    case (1, u) if u >= 1 => Teens(u - 1)
    case (t, 0) => Tens(t - 1)
    case (t, u) => Tens(t - 1) + "-" + Units(u - 1)
  }

  private def inEnglishSubThousand(number: Int): String = (number / 100, number % 100) match {
    case (0, subHundred) => inEnglishSubHundred(subHundred)
    case (hundreds, 0) => inEnglishSubHundred(hundreds) + " hundred"
    case (hundreds, subHundred) => inEnglishSubHundred(hundreds) + " hundred " + inEnglishSubHundred(subHundred)
  }

  private def inEnglishTriplet(number: Long, divisor: Int, tripletText: Option[String]): List[String] = {
    val subThousand = ((number / divisor) % 1000).toInt
    if (subThousand == 0) List()
    else if (tripletText.isEmpty) List(inEnglishSubThousand(subThousand))
    else List(inEnglishSubThousand(subThousand), tripletText.get)
  }

  def inEnglish(number: Long): Option[String] =
    if (number < 0 || number >= 1000000000000L) None
    else if (number == 0) Some("zero")
    else Some(Triplets.flatMap(e => inEnglishTriplet(number, e._1, e._2)).mkString(" "))
}